java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -242,21 +246,38 @@ public void visitLambdaExpression(LambdaExpressionTree lambdaExpressionTree) {
 
     @Override
     public void visitBinaryExpression(BinaryExpressionTree tree) {
-      if(tree.is(CONDITIONAL_AND, CONDITIONAL_OR)) {
-        increaseComplexityByOne(tree.operatorToken());
-        ExpressionTree left = ExpressionUtils.skipParentheses(tree.leftOperand());
-        if(left.kind() == tree.kind()) {
-          ignored.add(((BinaryExpressionTree) left).operatorToken());
-          complexity--;
-        }
-        ExpressionTree right = ExpressionUtils.skipParentheses(tree.rightOperand());
-        if(right.kind() == tree.kind()) {
-          ignored.add(((BinaryExpressionTree) right).operatorToken());
-          complexity--;
+      if (tree.is(CONDITIONAL_AND, CONDITIONAL_OR) && !ignored.contains(tree)) {
+        List<BinaryExpressionTree> flatOperators = flattenLogicalExpression(tree);
+
+        // first binary expression
+        increaseComplexityByOne(flatOperators.get(0).operatorToken());
+
+        // rest
+        for (int i = 1; i < flatOperators.size(); i++) {
+          BinaryExpressionTree previous = flatOperators.get(i - 1);
+          BinaryExpressionTree current = flatOperators.get(i);
+          if (!previous.is(current.kind())) {
+            increaseComplexityByOne(current.operatorToken());
+          }
         }
       }
       super.visitBinaryExpression(tree);
     }
+
+    private List<BinaryExpressionTree> flattenLogicalExpression(ExpressionTree expression) {
+      if (expression.is(CONDITIONAL_AND, CONDITIONAL_OR)) {
+        ignored.add(expression);
+
+        BinaryExpressionTree binaryExpr = (BinaryExpressionTree) expression;
+        ExpressionTree left = ExpressionUtils.skipParentheses(binaryExpr.leftOperand());
+        ExpressionTree right = ExpressionUtils.skipParentheses(binaryExpr.rightOperand());
+
+        return Stream.of(flattenLogicalExpression(left), Collections.singletonList(binaryExpr), flattenLogicalExpression(right))
+          .flatMap(List::stream)
+          .collect(Collectors.toList());
```

Comment:
you can most probably avoid collecting (and thus flatmapping) by relying on Stream.concat and Stream.empty and then collect on the resulting call.